### ISSUE TITLE
Alerting based on tracing data

### ISSUE THREAD

**Issue by muryoutaisuu** [jaegertracing/jaeger#1570]
## Requirement - what kind of business use case are you trying to solve?

We won't always have the dashboard opened to see whether some requests are taking too long. Hence make it possible, to alert when a request as a whole (not necessarily for every single service) exceeds a statically set threshold. For the customers, the threshold of the whole request matters, not that of the smaller services. This means that even if a Service A is doing really bad, it might be compensated by Service B, which is doing great.

Suppose a request with 3 Services A, B and C each with a set threshold, but the overall request time should be less than all 3 service thresholds together:

|                                                                                   | A     | B     | C    | Total                                                                   | Alert? |
|-----------------------------------------------------------------------------------|-------|-------|------|-------------------------------------------------------------------------|--------|
| Threshold                                                                         | 100ms | 200ms | 50ms | 350ms. Nevertheless, we want to achieve **300ms** over the whole request | -      |
| Usecase, where everything is ok                                                   | 70ms  | 130ms | 20ms | 220ms                                                                   | n      |
| Usecase, where one service and also the whole request exceeds threshold           | 200ms | 160ms | 40ms | 400ms                                                                   | y      |
| Usecase, where one service exceeds, but a different service compensates it        | 120ms | 50ms  | 30ms | 200ms                                                                   | n      |
| Usecase, where every Service is ok, but the whole request exceeds threshold | 90ms  | 180ms | 40ms | 310ms                                                                   | y      |

## Problem - what in Jaeger blocks you from solving the requirement?

Alerting is not implemented at all.

## Proposal - what do you suggest to solve the problem or improve the existing situation?

As for now I see two solutions with Prometheus & Alertmanager:

1. Implement metrics for each service, then create alertrules in Prometheus (where you calculate on your own, e.g. (pseudocode) timeA+timeB+timeC<300ms
2. Implement metrics for the whole request, then create alertrules in Prometheus on this single metric (without the need to sum up each service)

## Any open questions to address

1. Is it necessary to be able to alert on tracing data?
2. If yes, should it be done via Prometheus metrics or with its own implementation?
3. Should each Service or the whole request as a single metric be available for monitoring?

_small shoutout to @jpkrohling: Thank you for your talk and our little chat at kubecon2019 Barcelona! and for encouraging me to create this feature request_

**Comment by jpkrohling** [comment]
Thanks for opening this issue! While the alerting part should be handled somewhere else, I think there's value in having a "duration" metric per trace.

The tricky part is determining what's a trace, as we don't have such a definition right now.

One idea that would be right most of the time is to name traces based on the root span, with the assumption that all traces originating from the same place (ie, REST endpoint) are equivalent.

Most of it, or all of it, could probably be done already when using Kafka + ingester. We just need to build and document an example.

**Comment by objectiser** [comment]
One of the issues is knowing when a trace instance as completed - do we wait a certain time period without any more spans being reported, and assume it is complete? Do we have a "template" against which an instance can be compared? Perhaps the concept of completion is trace specific - so a metric is derived from the occurrence of a specific pair (or set) of spans in the trace instance?

**Comment by jpkrohling** [comment]
> One of the issues is knowing when a trace instance as completed

For this particular use case, we might not need it. We just need the trace ID + duration. Whenever a new span arrives for that trace, we just update the duration. Once the duration is higher than the threshold, an action is taken.

That said, it does sound like it belongs to a consumer of a stream of spans rather than Jaeger itself.

**Comment by yurishkuro** [comment]
> I think there's value in having a "duration" metric per trace.

@jpkrohling what do you mean by that? Just emitting a timer with service/operation as tags for all root spans? Why not _all spans_? And are you talking about the backend or client libs?

@muryoutaisuu there are many existing systems today that implement alerting, I don't know why we'd want to do that in Jaeger, unless there's something unique we can provide.

**Comment by jpkrohling** [comment]
Backend. Note that it's not a timer per span, but a timer named after the root span. The start time would be the start time of the root span and the duration would have to be calculated after pretty much every span in the trace. Does that make more sense now?

**Comment by muryoutaisuu** [comment]
@yurishkuro 
1. I don't feel I would get an accurate duration from other system like prometheus, which does not account network latency between the services.
2. it would be rather cumbersome to add all metric values together in one prometheus query, whereas it could be provided rather simply as an all-over-all metric in jaeger? *rather simply* being the view of the customer/enduser, not the ones doing the implementation.
3. if ad-hoc live tracing in the dashboard is the only value one gets out of implementing opentracing/jaeger in one's application, I feel it might not outweight the integration costs in most (smaller) companies. Hence a lower adoption-rate. _(No question, this point is very arguable and depends on companies budgets & resources)_

**Comment by yurishkuro** [comment]
I agree that emitting metrics from Jaeger backend, like latency of requests, could be valuable, especially if you sample all requests. What I'm not agreeing with is that we need to build _alerting_ capabilities on to of those metrics. Eg you could scrape those metrics from Jaeger collectors using Prometheus and alert there. You can even use trace data to pregenerate the alert rules.

**Comment by muryoutaisuu** [comment]
@yurishkuro I probably didn't emphasis it enough that I agree on alerting via Prometheus & Alertmanager as well :sweat_smile:
I'll rename this issue to sth. else, removing alerting from the title.

**Comment by annanay25** [comment]
Thanks for reporting this issue, @muryoutaisuu. I agree that there is value in reporting latency metrics at the backend. 

> Implement metrics for each service, then create alertrules in Prometheus (where you calculate on your own, e.g. (pseudocode) timeA+timeB+timeC<300ms (actually this might be implemented with Prometheus only, no need for Jaeger)

+1 for this approach.
Pros -  
- Exposing latency metrics per service from Jaeger and defining alert rules on alertmanager seems like a clean solution. Tracking the service-path for full-trace metrics could get tricky.
- Exposing metrics per path could bloat the number of metrics.

Cons -
- A possible corner-case would be if a service performs different computations for requests from different sources, in which case the latency metrics would be contorted. 
- If the service graph is not deterministic (requests could take multiple paths based on parameters), alert rules would be hard to define.

**Comment by yurishkuro** [comment]
@annanay25 your PR only works if people are using 100% sampling. If they use probabilistic sampling, then the numbers / histogram are going to be completely wrong & misleading. It is possible to make the calculations more accurate by taking into account the sampling rate (the extrapolation will be "correct" but subject to statistical sample size error margins), but we currently do not capture/propagate sampling probability in the trace context, thus the correct measurements can only be calculated by having access to the root span of the trace.

**Comment by annanay25** [comment]
Ah, yes. Thanks for the observations. In that case, I believe naming the metrics after the root spans would be the best way to go about this. How would we group different trace-ids into a common alert (based on the path) then? 

@jpkrohling suggested the following - 
> One idea that would be right most of the time is to name traces based on the root span, with the assumption that all traces originating from the same place (ie, REST endpoint) are equivalent.

**Comment by yurishkuro** [comment]
That is the main challenge. Collectors are stateless and never see the full trace, only one span at a time.

**Comment by muryoutaisuu** [comment]
I just talked with a colleague. If the first hit service correctly emits duration metrics for Prometheus, there's no need for Jaeger to emit trace duration metrics. One can make alerts on those first service metrics in Prometheus and then analyse the issue with the Jaeger GUI.

So once again I'm back to being unsure whether this feature is needed and reasonable inside Jaeger. I'm sorry for making a fuss 🙈

**Comment by jpkrohling** [comment]
> If the first hit service correctly emits duration metrics for Prometheus, there's no need for Jaeger to emit trace duration metrics

This is true only if all the calls are synchronous. If you have async calls, your trace duration isn't the same as the root span's duration.
